applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/table/TableCellComponent.java

```diff
@@ -16,39 +16,135 @@
 import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
 import org.apache.sling.api.resource.Resource;
-import org.apache.sling.models.annotations.Default;
 import org.apache.sling.models.annotations.DefaultInjectionStrategy;
 import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.injectorspecific.SlingObject;
+import pl.ds.kyanite.common.components.models.table.dto.TableCellData;
+import pl.ds.kyanite.common.components.models.table.dto.TableRowData;
 
+@Slf4j
 @Model(adaptables = Resource.class, defaultInjectionStrategy = DefaultInjectionStrategy.OPTIONAL)
 public class TableCellComponent {
 
   private static final String DEFAULT_TEXT = "Content";
+  private static final int DEFAULT_COLSPAN = 1;
+  private static final int DEFAULT_ROWSPAN = 1;
 
   @Inject
   @Getter
-  @Default(values = DEFAULT_TEXT)
   private String text;
 
   @Inject
-  @Getter
-  @Default(intValues = 1)
-  private int columns;
+  private Integer columns;
 
   @Inject
-  @Getter
-  @Default(intValues = 1)
-  private int rows;
+  private Integer rows;
+
+  @SlingObject
+  private Resource currentResource;
 
   @PostConstruct
   private void init() {
-    if (text.isEmpty()) {
+    if (StringUtils.isEmpty(text)) {
       text = DEFAULT_TEXT;
     }
   }
 
+  public int getColspan() {
+    return columns != null ? columns : computeColspan();
+  }
+
+  public int getRowspan() {
+    return rows != null ? rows : DEFAULT_ROWSPAN;
+  }
+
+  private int computeColspan() {
+    Optional<TableComponent> parentTable = Optional
+            .ofNullable(currentResource.getParent()) // table row
+            .map(Resource::getParent) // table head or body
+            .map(Resource::getParent) // the table
+            .map(tableResource -> tableResource.adaptTo(TableComponent.class));
+
+    if (parentTable.isEmpty() || parentTable.get().isScrollable()) {
+      return DEFAULT_COLSPAN;
+    }
+
+    Resource parentRow = currentResource.getParent();
+
+    List<TableCellData> tableCells = StreamSupport
+            .stream(parentRow.getChildren().spliterator(), false)
+            .map(cell -> new TableCellData(
+                    cell.getValueMap().get("columns", Integer.class),
+                    cell.getPath().equals(currentResource.getPath())))
+            .toList();
+
+    TableRowData tableRowData = new TableRowData(tableCells);
+
+    return computeColspan(tableRowData);
+  }
+
+  /**
+   * Computes colspan for a cell that hasn't got the columns property specified.
+   * Attempts to give each such cell the same colspan, if possible.
+   * It takes into account colspans from other cells in the same row.
+   */
+  private static int computeColspan(TableRowData row) {
+    int summaryAvailableColspan = TableComponent.DEFAULT_SUMMARY_COLSPAN
+            - row.getSummaryColspanTakenByCellsWithColumnsSpecified();
+
+    if (summaryAvailableColspan < 1) {
+      return 1;
+    }
```

Comment:
This can happen for table `|6| |6|` - when user adds a cell between two cells with explicit columns=6.
We still want the cell to be visible. So this table row would grow to 13 "colspan points" in this case